feat: Optionally retain external image URLs identified by specified url prefix

sdmcraft opened the pull request:
Please ensure your pull request adheres to the following guidelines:
- [ ] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues
#724

tripodsan commented on src/mdast-process-images.js:
```diff
@@ -44,17 +64,31 @@ export async function processImages(log, tree, mediaHandler, baseUrl) {
         node.url = new URL(url, baseUrl).href;
         register(node);
       } else if (url.startsWith('https://')) {
+        if (isExternalImage(url, externalImages)) {
+          // Add custom class to the node for external images
+          const data = node.data || {};
+          const hProperties = data.hProperties || {};
+          hProperties.className = 'external-image';
+          // eslint-disable-next-line no-param-reassign
+          node.data = { ...data, hProperties };
+
+          // eslint-disable-next-line no-param-reassign
+          node.data.externalImage = true;
+
+          log.debug(`Marked external image: ${url}`);
+        }
+
```
why do you need to check it here and not just in `register()` ?

sdmcraft replied:
Agreed. Simplified.

tripodsan commented on src/mdast-process-images.js:
```diff
@@ -44,17 +64,31 @@ export async function processImages(log, tree, mediaHandler, baseUrl) {
 
-  if (images.size > 200) {
-    throw new TooManyImagesError(`maximum number of images reached: ${images.size} of 200 max.`);
+  if (images.size + externalImageNodes.size > 200) {
+    throw new TooManyImagesError(`maximum number of images reached: ${images.size + externalImageNodes.size} of 200 max.`);
```
we don't care if the document has too many external images. no need to include them in the check.

sdmcraft replied:
done

tripodsan commented on src/index.js:
```diff
@@ -200,6 +200,7 @@ async function run(request, ctx) {
       org,
       site,
       unspreadLists: !!ctx.data.features?.unspreadLists,
+      externalImagesUrlPrefixes: ctx.data.features?.externalImagesUrlPrefix,
```
the flag is currently called `externalImageUrlPrefix`

sdmcraft commented:
@tripodsan , thanks for reviewing. I believe I have addressed all the review feedback. Also tested it with the deployed ci build. Marking this ready to merge. Could you please take a look and let me know if it good to merge?

tripodsan commented on src/html2md.js:
```diff
@@ -338,7 +338,7 @@ export async function html2md(html, opts) {
   addMetadata(hast, mdast);
 
-  await processImages(log, mdast, mediaHandler, url);
+  await processImages(log, mdast, mediaHandler, url, opts.externalImagesUrlPrefixes);
```
```suggestion
  await processImages(log, mdast, mediaHandler, url, opts.externalImageUrlPrefixes);
```

tripodsan commented on src/index.js:
```diff
@@ -200,6 +200,7 @@ async function run(request, ctx) {
       org,
       site,
       unspreadLists: !!ctx.data.features?.unspreadLists,
+      externalImagesUrlPrefixes: ctx.data.features?.externalImageUrlPrefixes,
```
```suggestion
      externalImageUrlPrefixes: ctx.data.features?.externalImageUrlPrefixes,
```

tripodsan commented on src/mdast-process-images.js:
```diff
@@ -21,14 +21,43 @@ export class TooManyImagesError extends Error {
  * @param {object} tree
  * @param {MediaHandler} mediaHandler
  * @param {string} baseUrl
+ * @param {Array<string>} externalImagesUrlPrefixes Array of url prefixes to detect external images
```
```suggestion
 * @param {Array<string>} externalImageUrlPrefixes Array of url prefixes to detect external images
```

tripodsan commented on src/mdast-process-images.js:
```diff
@@ -21,14 +21,43 @@ export class TooManyImagesError extends Error {
+  tree,
+  mediaHandler,
+  baseUrl,
+  externalImagesUrlPrefixes = [],
```
```suggestion
  externalImageUrlPrefixes = [],
```

tripodsan commented on src/mdast-process-images.js:
```diff
@@ -21,14 +21,43 @@ export class TooManyImagesError extends Error {
   }
   // gather all image nodes
   const images = new Map();
+  const externalImageNodes = new Map();
```
why do you track the external images, when you don't use the collection afterwards?

sdmcraft replied:
i think its a remnant of my earlier change wherein I was counting the external images for the 200 limit. Since we no longer do that, no need for this tracking as well. Removing.

tripodsan reviewed (changes requested):
no need to track the external images

tripodsan commented on src/mdast-process-images.js:
```diff
@@ -21,14 +21,37 @@ export class TooManyImagesError extends Error {
+
   const register = (node) => {
+    // Check if this is an external image
+    const { url = '' } = node;
```
you extract url here, but then don't use it later (line 55/56)

sdmcraft replied:
yes, that was pre-existing but good to use `url` there as well.

tripodsan commented on test/mdast-process-images.test.js:
```diff
@@ -0,0 +1,779 @@
+/*
+ * Copyright 2022 Adobe. All rights reserved.
```
```suggestion
 * Copyright 2025 Adobe. All rights reserved.
```

tripodsan commented on test/mdast-process-images.test.js:
```diff
@@ -0,0 +1,779 @@
+    assert.strictEqual(processedUrls[0], 'https://regular-image.com/image.jpg', 'Regular image should be processed');
+  });
+
+  it('handles duplicate external asset URLs', async () => {
```
this test is not really needed anymore
